File: 10_syntax_scoring/syntax_scoring.py
import statistics
from typing import List

SET_NOTATIONS = set([')', ']', '}', '>'])
NOTATION_MAPPING = {
    ')': '(',
    ']': '[',
    '}': '{',
    '>': '<',
    '(': ')',
    '[': ']',
    '{': '}',
    '<': '>',
}
WRONG_NOTATION_SCORE = {
    ')': 3,
    ']': 57,
    '}': 1197,
    '>': 25137
}
COMPLETE_SCORE = {
    ')': 1,
    ']': 2,
    '}': 3,
    '>': 4
}


def get_inputs(file_name: str) -> List[str]:
    with open(file_name, 'r') as f:
        lines = f.readlines()
    return [line.strip() for line in lines]


def syntax_error_score(lines: List[str]) -> int:
    score = 0
    for line in lines:
        stack = []
        for char in line:
            if char == '(' or char == '[' or char == '{' or char == '<':
                stack.append(char)
            else:
                if len(stack) > 0 and stack[-1] == NOTATION_MAPPING[char]:
                    stack.pop()
                else:
                    stack.append(char)
        if len(stack) == 0:
            continue
        set_stack = set(stack)
        if len(set_stack.intersection(SET_NOTATIONS)) == 0:
            continue
        wrong_notations = list(set_stack.intersection(SET_NOTATIONS))
        min_idx = len(stack)
        wrong_notation = wrong_notations[0]
        for notation in wrong_notations:
            idx = stack.index(notation)
            if idx < min_idx:
                min_idx = idx
                wrong_notation = notation
        score += WRONG_NOTATION_SCORE[wrong_notation]
    return score


def fix_incomplete(lines: List[str]) -> int:
    scores = []
    for line in lines:
        stack = []
        for char in line:
            if char == '(' or char == '[' or char == '{' or char == '<':
                stack.append(char)
            else:
                if len(stack) > 0 and stack[-1] == NOTATION_MAPPING[char]:
                    stack.pop()
                else:
                    stack.append(char)
        if len(stack) == 0:
            continue
        set_stack = set(stack)
        if len(set_stack.intersection(SET_NOTATIONS)) != 0:
            continue
        addition = []
        for i in range(len(stack)-1, -1, -1):
            char = stack[i]
            addition.append(NOTATION_MAPPING[char])
        score = 0
        for char in addition:
            score *= 5
            score += COMPLETE_SCORE[char]
        scores.append(score)
    return int(statistics.median(scores))


if __name__ == '__main__':
    data = get_inputs('input.txt')
    print(f'Err score: {syntax_error_score(data)}')
    print(f'Incomplete score: {fix_incomplete(data)}')

File: 10_syntax_scoring/syntax_scoring.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strings"
)

var setNotations = []rune{')', ']', '}', '>'}
var notationMapping = map[rune]rune{
	'(': ')',
	'[': ']',
	'{': '}',
	'<': '>',
	')': '(',
	']': '[',
	'}': '{',
	'>': '<',
}
var wrongNotationScoreMapping = map[rune]int{
	')': 3,
	']': 57,
	'}': 1197,
	'>': 25137,
}
var completeScoreMapping = map[rune]int{
	')': 1,
	']': 2,
	'}': 3,
	'>': 4,
}

func getInputs(filename string) ([]string, error) {
	data, err := os.ReadFile(filename)
	if err != nil {
		return nil, err
	}
	return strings.Split(string(data), "\n"), nil
}

func intersection(a, b []rune) []rune {
	m := make(map[rune]bool)
	for _, v := range a {
		m[v] = true
	}
	var result []rune
	for _, v := range b {
		if checked, ok := m[v]; ok {
			if checked {
				result = append(result, v)
				m[v] = false
			}
		}
	}
	return result
}

func syntaxErrScore(lines []string) int {
	var score int
	for _, line := range lines {
		var stack []rune
		for _, char := range line {
			if char == '(' || char == '[' || char == '{' || char == '<' {
				stack = append(stack, char)
				continue
			}
			if len(stack) > 0 && stack[len(stack)-1] == notationMapping[char] {
				stack = stack[:len(stack)-1]
			} else {
				stack = append(stack, char)
			}
		}
		if len(stack) == 0 {
			continue
		}
		wrongNotations := intersection(stack, setNotations)
		if len(wrongNotations) == 0 {
			continue
		}
		var wrongNotation rune
		minIdx := len(stack)
		for _, notation := range wrongNotations {
			if idx := strings.IndexRune(string(stack), notation); idx < minIdx {
				fmt.Println(idx)
				minIdx = idx
				wrongNotation = notation
			}
		}
		score += wrongNotationScoreMapping[wrongNotation]
	}
	return score
}

func fixIncomplete(lines []string) int {
	var scores []int
	for _, line := range lines {
		var stack []rune
		for _, char := range line {
			if char == '(' || char == '[' || char == '{' || char == '<' {
				stack = append(stack, char)
				continue
			}
			if len(stack) > 0 && stack[len(stack)-1] == notationMapping[char] {
				stack = stack[:len(stack)-1]
			} else {
				stack = append(stack, char)
			}
		}
		if len(stack) == 0 {
			continue
		}
		wrongNotations := intersection(stack, setNotations)
		if len(wrongNotations) != 0 {
			continue
		}
		var addition []rune
		for i := len(stack) - 1; i >= 0; i-- {
			addition = append(addition, rune(notationMapping[stack[i]]))
		}
		var score int
		for _, char := range addition {
			score *= 5
			score += completeScoreMapping[char]
		}
		scores = append(scores, score)
	}
	sort.Ints(scores)
	if len(scores)%2 == 0 {
		return (scores[len(scores)/2] + scores[len(scores)/2-1]) / 2
	}
	return scores[len(scores)/2]
}

func main() {
	inputs, err := getInputs("input.txt")
	if err != nil {
		fmt.Println(err)
		return
	}
	fmt.Printf("Err score: %d\n", syntaxErrScore(inputs))
	fmt.Println("Incomplete score:", fixIncomplete(inputs))
}
